java/connector-node/risingwave-connector-service/src/main/java/com/risingwave/connector/source/common/PostgresValidator.java

```diff
@@ -384,17 +388,91 @@ private void validatePublicationConfig(boolean isSuperUser) throws SQLException
-            // make sure the publication are created with `publish_via_partition_root = true`, which
-            // is required by partitioned tables.
-            throw ValidatorUtils.invalidArgument(
-                    "Table '"
-                            + tableName
-                            + "' has partitions, which requires publication '"
-                            + pubName
-                            + "' to be created with `publish_via_partition_root = true`. \nHint: you can run `SELECT pubviaroot from pg_publication WHERE pubname = '"
-                            + pubName
-                            + "'` in the upstream Postgres to check.");
+
+        if (!partitions.isEmpty() && isPublicationExists) {
+            // `pubviaroot` in `pg_publication` is added after PG v13, before which PG does not
+            // allow adding partitioned table to a publication. So here, if partitions.isEmpty() is
+            // false, we can safely check the value of `pubviaroot` of the publication here.
+            boolean isPublicationViaRoot = false;
+            try (var stmt =
+                    jdbcConnection.prepareStatement(
+                            ValidatorUtils.getSql("postgres.publication_pubviaroot"))) {
+                stmt.setString(1, pubName);
+                var res = stmt.executeQuery();
+                if (res.next()) {
+                    isPublicationViaRoot = res.getBoolean(1);
+                }
+            }
+            if (!isPublicationViaRoot) {
+                // Make sure the publication are created with `publish_via_partition_root = true`,
+                // which is required by partitioned tables.
+                throw ValidatorUtils.invalidArgument(
+                        "Table '"
+                                + tableName
+                                + "' has partitions, which requires publication '"
+                                + pubName
+                                + "' to be created with `publish_via_partition_root = true`. \nHint: you can run `SELECT pubviaroot from pg_publication WHERE pubname = '"
+                                + pubName
+                                + "'` in the upstream Postgres to check.");
+            }
+        }
+        //  Only after v13, PG allows adding a partitioned table to a publication. So, if the
+        // version is before v13, the tables in a publication are always partition leaves, we don't
+        // check their ancestors and descendants anymore.
+        if (isPublicationExists && pgVersion >= 13) {
+            List<String> family = new ArrayList<>();
+            boolean findRoot = false;
+            String currentPartition = tableName;
+            while (!findRoot) {
+                try (var stmt =
+                        jdbcConnection.prepareStatement(
+                                ValidatorUtils.getSql("postgres.partition_parent"))) {
```

Comment:
The output of `pg_partition_tree` has a `level` column ([ref](https://pgpedia.info/p/pg_partition_tree.html)), so we can easily check if the `schemaName.tableName` is a root table. I think we should only allow user to `CREATE TABLE` for the root table, and don't need to check whether descendant of the root is already covered in the publication.

Reply:
The `level` here is actually relative level instead of absolute level. See [here](https://www.postgresql.org/docs/current/functions-admin.html#:~:text=The%20level%20value%20is%200%20for%20the%20input%20table%20or%20index). Checking descendant is more trivial on the contrary.

Reply:
> The level here is actually relative level instead of absolute level.

You are right.